src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt

```diff
@@ -391,41 +434,47 @@ class CosmasGoogleCloudService(private val freeBucketName: String,
         val project = synchronized(this.fileBuffer) {
             this.fileBuffer.getValue(request.info.projectId)
         }
-        // get last version from storage
-        val blob: Blob? = try {
+        // get last version from storage or default instance if it doesn't exist
+        val latestVersionBlob: Blob? = try {
             this.storage.get(getBlobId(request.fileId, request.info))
         } catch (e: StorageException) {
             handleStorageException(e, responseObserver)
             return
         }
-        val previousVersion = if (blob != null) {
-            CosmasProto.FileVersion.parseFrom(blob.getContent()).content.toStringUtf8()
-        } else ""
-
-        val actualVersion = request.actualContent.toStringUtf8()
-        val patch = getDiffPatch(previousVersion, actualVersion, request.timestamp)
-
-        val newVersion = CosmasProto.FileVersion.newBuilder()
+        val latestVersion = if (latestVersionBlob != null) {
+            CosmasProto.FileVersion.parseFrom(latestVersionBlob.getContent())
+        } else FileVersion.getDefaultInstance()
+
+        // Preparing new version in memory to replace bad or nonexistent one in buffer
+        // Window should point to the latest N versions
+        val windowToCommit = if (latestVersionBlob != null) {
+            val latestVersionInfo = FileVersionInfo.newBuilder()
+                    .setFileId(request.fileId)
+                    // For in-memory storage implementation resultBlob.generation == null,
+                    // but in this case we don't care about generation value, so I set it to 1L
+                    .setGeneration(latestVersionBlob.generation ?: 1L)
+                    .setTimestamp(latestVersion.timestamp)
+                    .build()
+            getNewWindow(latestVersionInfo, latestVersion.historyWindowList)
+        } else emptyList<FileVersionInfo>()
+        // Content is taken from the latest versions
```

Comment:
This comment is misleading, because "latestVersion" is already used for the latest stored version. We, however, do not take the contents from there.